lift/src/main/scala/com/linkedin/lift/mitigation/EOppUtils.scala

```diff
@@ -0,0 +1,156 @@
+package com.linkedin.lift.mitigation
+
+import com.linkedin.lift.types.{ScoreWithAttribute, ScoreWithLabelAndAttribute}
+import org.apache.spark.sql.{DataFrame, Dataset, SparkSession}
+
+
+object EOppUtils {
+
+
+  /**
+    * Transforming a single score using a transformation function given as a scala map.
+    * We first perform a binary search to determine the closest lowerBound and upperBound of the score in the keys
+    * of the transformation map. Then we transform the score assuming that the transformation function is linear in
+    * the interval (lowerBound, upperBound).
+    *
+    * @param score          score value
+    * @param sortedKeys     sorted keys of the transformation map
+    * @param transformation transformation function given as a scala map
+    * @return transformed score
+    */
+  def transformScore(score: Double, sortedKeys: List[Double], transformation: Map[Double, Double]): Double = {
```

Comment:
I feel we don't need the keys be sorted, since that requirement results in an O(nlogn) precomputation (prior to calling this method) + O(logn) search (in this method). Rather, we can just have this method find the bounds in O(n), and it might simplify the logic in this method.

Scala Sequences have the `indexWhere` and `lastIndexWhere` methods that allows you to find the first and last indices where a predicate is satisfied. The `min` and `max` methods for the sequence will find the `head` and `last` values of a sorted sequence.

Reply:
This is a helper function for applyTransformation() which transforms $m$ scores based on a Map with $n$ keys. Using the sortedKeys let us do it in O(nlogn + mlogn). Without using the sorted keys, it would be O(mn).

Does it make sense now?

Reply:
Sounds good, thanks!